[SVG] Clean up SVGTextMetricsBuilder

Constellation opened the pull request:
#### d0ecc42ca2afd958eb7831d67e5bdb19c48ab007
<pre>
[SVG] Clean up SVGTextMetricsBuilder
<a href="https://bugs.webkit.org/show_bug.cgi?id=271717">https://bugs.webkit.org/show_bug.cgi?id=271717</a>
<a href="https://rdar.apple.com/125428940">rdar://125428940</a>

Reviewed by Cameron McCormack.

This patch adds optimized fast path for SVGTextMetricsBuilder.

1. We reorganized RenderSVGText::layout so that we no longer call SVGTextMetricsBuilder multiple times
   unnecessarily in one layout. We should call once per layout.
2. We use canUseSimplifiedTextMeasuring information from RenderSVGInlineText (this flag gets updated correctly based on style change etc.).
   And based on that, we use super tuned fast path for SVGTextMetrics collection. We use FontCascade&apos;s WidthCache well. And we skip various
   unnecessarily costly operation done for per character basis on the normal path.
3. Clean up SVGTextMetricsBuilder code. Make MeasureTextData const parameter for SVGTextMetricsBuilder::measureTextRenderer to make input
   and output much more explicit and stateless. Remove many unnecessary branches.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleTextWithFixedPitch const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::widthForTextUsingSimplifiedMeasuring const):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::updateFontInAllDescendants):
(WebCore::RenderSVGText::layout):
(WebCore::RenderSVGText::willLayout): Deleted.
* Source/WebCore/rendering/svg/RenderSVGText.h:
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer):
(WebCore::MeasureTextData::MeasureTextData):
(WebCore::SVGTextMetricsBuilder::measureTextRenderer):
(WebCore::SVGTextMetricsBuilder::walkTree):
(WebCore::SVGTextMetricsBuilder::buildMetricsAndLayoutAttributes):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.h:

Canonical link: <a href="https://commits.webkit.org/277214@main">https://commits.webkit.org/277214@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5554e58a6012aa12dcd391d6b9d91f48d6a91003

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49652 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49698 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43063 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49323 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/31336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23651 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47597 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/31336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/40508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/19599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/31336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5061 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/31336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/42048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51574 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/18397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/45585 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/23317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/44579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6595 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/23028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -578,11 +578,10 @@ static bool canUseSimplifiedTextMeasuringForCharacters(std::span<const Character
 }
 
-bool TextUtil::canUseSimplifiedTextMeasuring(StringView textContent, const RenderStyle& style, const RenderStyle* firstLineStyle)
+bool TextUtil::canUseSimplifiedTextMeasuring(StringView textContent, const FontCascade& fontCascade, bool whitespaceIsCollapsed, const RenderStyle* firstLineStyle)
```
I have a slight preference to passing in a `WhiteSpace` value here instead of a `bool`, and calling `RenderStyle::collapseWhiteSpace(theValue)` below. The current call sites are relatively clear, but it's good to avoid bool-typed arguments IMO.

Constellation replied:
Right now, `WidthIterator::characterCanUseSimplifiedTextMeasuring` is taking bool, and this is a bit important since `FontCascade::canUseSimplifiedTextMeasuring` is always passing `true` for that since it already checked for `false` case before calling `WidthIterator::characterCanUseSimplifiedTextMeasuring`, and FontCascade would like to cache values only for `whitespaceIsCollapsed = true`. For now, I'll keep it as is since all the other places are using `bool` for `whitespaceIsCollapsed` parameter in width computation. We could consider refactoring this later.

heycam commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -102,88 +102,126 @@ void SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer(RenderSVGInlin
+                }
+
+                if ((valueListPosition + i - skippedCharacters + 1) == defaultPosition)
+                    attributes->characterDataMap().set(i + 1, WTFMove(characterData));
```
At first I wondered whether we could move from `characterData` more than once in this loop, but we can't, so it's OK. Though `SVGCharacterData` is just a struct of floats, so doesn't benefit from a `WTFMove` anyway.

Constellation replied:
Yeah, without WTFMove is fine.

heycam commented on Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
```diff
@@ -102,88 +102,126 @@ void SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer(RenderSVGInlin
+            ASSERT(scalingFactor);
+            float scaledHeight = scaledFont.metricsOfPrimaryFont().height() / scalingFactor;
+
+            // m_canUseSimplifiedTextMeasuring ensures that this does not include surrogate pairs. So we do not need to consider about them.
```
Can we assert this in the loop?

Constellation replied:
Added.

Constellation committed:
[SVG] Clean up SVGTextMetricsBuilder
https://bugs.webkit.org/show_bug.cgi?id=271717
rdar://125428940

Reviewed by Cameron McCormack.

This patch adds optimized fast path for SVGTextMetricsBuilder.

1. We reorganized RenderSVGText::layout so that we no longer call SVGTextMetricsBuilder multiple times
   unnecessarily in one layout. We should call once per layout.
2. We use canUseSimplifiedTextMeasuring information from RenderSVGInlineText (this flag gets updated correctly based on style change etc.).
   And based on that, we use super tuned fast path for SVGTextMetrics collection. We use FontCascade's WidthCache well. And we skip various
   unnecessarily costly operation done for per character basis on the normal path.
3. Clean up SVGTextMetricsBuilder code. Make MeasureTextData const parameter for SVGTextMetricsBuilder::measureTextRenderer to make input
   and output much more explicit and stateless. Remove many unnecessary branches.

* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::widthForSimpleTextWithFixedPitch const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::widthForTextUsingSimplifiedMeasuring const):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::updateFontInAllDescendants):
(WebCore::RenderSVGText::layout):
(WebCore::RenderSVGText::willLayout): Deleted.
* Source/WebCore/rendering/svg/RenderSVGText.h:
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.cpp:
(WebCore::SVGTextMetricsBuilder::initializeMeasurementWithTextRenderer):
(WebCore::MeasureTextData::MeasureTextData):
(WebCore::SVGTextMetricsBuilder::measureTextRenderer):
(WebCore::SVGTextMetricsBuilder::walkTree):
(WebCore::SVGTextMetricsBuilder::buildMetricsAndLayoutAttributes):
* Source/WebCore/rendering/svg/SVGTextMetricsBuilder.h:

Canonical link: https://commits.webkit.org/277214@main
